crates/server/src/admin/utils/auth.rs

```diff
@@ -283,6 +284,93 @@ pub fn validate_root_key_exists(
+                        return Err(ApiError {
+                            status_code: StatusCode::BAD_REQUEST,
+                            message: "Root key does not exist".into(),
+                        });
+                    }
+                    Ok(keys) => keys,
+                    Err(err) => {
+                        info!("Error checking if near client key exists: {}", err);
+                        return Err(ApiError {
+                            status_code: StatusCode::INTERNAL_SERVER_ERROR,
+                            message: err.to_string(),
+                        });
+                    }
+                };
+
+                // Extract wallet_address as a &str
+                let wallet_address = match req.wallet_metadata.wallet_address.as_deref() {
+                    Some(address) => address,
+                    None => {
+                        return Err(ApiError {
+                            status_code: StatusCode::BAD_REQUEST,
+                            message: "Wallet address not present".to_string(),
+                        });
+                    }
+                };
+
+                // Get network_type and from it use correct rpc_url
+                let rpc_url = match network_id {
+                    NearNetworkId::Mainnet => "https://rpc.mainnet.near.org",
+                    NearNetworkId::Testnet => "https://rpc.testnet.near.org",
+                    _ => {
+                        // Handle the case where the network ID is unknown or not handled
+                        return Err(ApiError {
+                            status_code: StatusCode::BAD_REQUEST,
+                            message: "Unknown NEAR network ID".into(),
+                        });
+                    }
+                };
+
+                // Check if the given public key is from the given NEAR account
+                if !check_for_near_account_key(
+                    &req.wallet_metadata.verifying_key,
+                    wallet_address,
+                    rpc_url,
+                )
+                .await?
+                {
+                    return Err(ApiError {
+                        status_code: StatusCode::BAD_REQUEST,
+                        message: "Given public key is not from wallet".into(),
+                    });
+                }
+                // Check if the wallet_address has a NEAR account key from DB
+                match check_for_near_account_key(&near_keys, wallet_address, rpc_url).await? {
+                    true => {
+                        let _ = store_root_key(
+                            req.wallet_metadata.verifying_key.clone(),
+                            req.wallet_metadata.wallet_type.clone(),
+                            wallet_address.to_string(),
+                            store,
+                        )
+                        .map_err(|err| {
+                            return err;
+                        })?;
```

Comment:
This `map_err` is unnecessary - the `return` returns from the map closure, so that's the same thing as just doing `err`, and as `err` was passed in, nothing changes 🙂 You can just remove it, and keep the `?`.